Implemented CategoricalGibbsMetropolis, optimized BinaryGibbsMetropolis.

ozankabak opened the pull request:
I took up on the discussion we had in Issue #1431, and implemented the `CategoricalGibbsMetropolis` sampler @twiecki and I contemplated there. I also optimized the implementation of `BinaryGibbsMetropolis`, and deprecated `ElemwiseCategorical`.

ozankabak committed:
Implemented CategoricalGibbsMetropolis, optimized BinaryGibbsMetropolis.

springcoil commented:
I think this is fine based on what @twiecki said.

On 11 Oct 2016 4:54 PM, "Mehmet Ozan Kabak" notifications@github.com
wrote:

> I took up on the discussion we had in Issue #1431
> https://github.com/pymc-devs/pymc3/issues/1431, and implemented the
> CategoricalGibbsMetropolis sampler @twiecki https://github.com/twiecki
> and I contemplated there. I also optimized the implementation of
> 
> ## BinaryGibbsMetropolis, and deprecated ElemwiseCategorical.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/pymc-devs/pymc3/pull/1439
> Commit Summary
> - Implemented CategoricalGibbsMetropolis, optimized
>   BinaryGibbsMetropolis.
> 
> File Changes
> - _M_ pymc3/sampling.py
>   https://github.com/pymc-devs/pymc3/pull/1439/files#diff-0 (5)
> - _M_ pymc3/step_methods/**init**.py
>   https://github.com/pymc-devs/pymc3/pull/1439/files#diff-1 (1)
> - _M_ pymc3/step_methods/gibbs.py
>   https://github.com/pymc-devs/pymc3/pull/1439/files#diff-2 (8)
> - _M_ pymc3/step_methods/metropolis.py
>   https://github.com/pymc-devs/pymc3/pull/1439/files#diff-3 (156)
> 
> Patch Links:
> - https://github.com/pymc-devs/pymc3/pull/1439.patch
> - https://github.com/pymc-devs/pymc3/pull/1439.diff
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/1439, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AA8DiB8SsFpdADalEZVhI8VQ9ZOVzqyBks5qy7EtgaJpZM4KTzD0
> .

fonnesbeck commented:
There should also be one or more unit tests for the step method.

twiecki commented on pymc3/step_methods/metropolis.py:
```diff
@@ -283,6 +295,122 @@ def competence(var):
+        candidate += 1
+    return candidate
+
+def softmax(x):
```
Why not use `theano.tensor.nnet.softmax`?

twiecki replied:
Can you use this version for more numerical stability: http://www.johndcook.com/blog/2010/01/20/how-to-compute-the-soft-maximum/ ?

ozankabak replied:
I am already making use of that trick. See the `- np.max(x)` term there? This multivariate form is a generalization of John Cook's bivariate expression.

twiecki commented on pymc3/step_methods/metropolis.py:
```diff
@@ -283,6 +295,122 @@ def competence(var):
         return Competence.INCOMPATIBLE
 
+class CategoricalGibbsMetropolis(ArrayStep):
+    """Metropolis-Hastings optimized for categorical variables. This step
```
Also add info that this is doing "Gibbs" -- i.e. updating each dimension one at a time, rather than blocked.

ozankabak replied:
Done. See new commit.

twiecki commented:
Agreed on tests. 

@wanderer2 Why is the binary step method more efficient? Or rather, why can't the Categorical be also as efficient?

ozankabak commented:
OK, I will send one more commit with:
1. A unit test for the `CategoricalGibbsMetropolis` class.
2. Add a note about the "Gibbs" aspect to both `BinaryGibbsMetropolis` and `CategoricalGibbsMetropolis` classes.

@twiecki: Theano's `nnet.softmax` is basically equivalent to the two-liner I used. See [here](http://deeplearning.net/software/theano/library/tensor/nnet/nnet.html#theano.tensor.nnet.nnet.softmax). In this case, I don't think there is a significant advantage to using Theano's softmax, but we can use it instead if you guys do.

About the binary step method: It is more efficient because it proposes new samples by simple binary negation, whereas the categorical step method would incur the overhead of going through random sampling only to propose the same value.

ozankabak committed:
Added unit test for CategoricalGibbsMetropolis.
Also Added extra comments requested by PyMC3 devs, and fixed a minor bug
in the proportional proposal utilized by CategoricalGibbsMetropolis.

ozankabak commented:
Hi guys. This latest commit has the additions you guys requested. Cheers!

ozankabak committed:
Fix a minor bug in the newly added CategoricalGibbsMetropolis test.

twiecki commented:
This looks good. Can you check the example notebooks whether they need updating?

ColCarroll commented:
I have been working on updating the example notebooks  -- a bunch have drifted out of date.  I'll  push what I have later tonight.

ozankabak commented:
Looking at the notebooks after @ColCarroll's updates, the only notebook that uses `ElemwiseCategorical`  is `gaussian_mixture_model`. We can change it to use `CategoricalGibbsMetropolis` right away, or make the change later on when/if we decide to remove `ElemwiseCategorical` completely. (My commits only deprecate `ElemwiseCategorical`; it is not removed.)

I say we keep that notebook as is (for now) and update it later on. It will provide an example use for `ElemwiseCategorical` while it is still a part of the library.

As a bonus, I have been working on a new notebook that demonstrates how `CategoricalGibbsMetropolis` is used. It is still a WIP, and it might take a while before I create a PR for it.

All in all, this PR is ready to get merged, in my opinion. Any objections?

twiecki commented:
Thanks for the contribution!
